components/service/glean/src/main/java/mozilla/components/service/glean/scheduler/MetricsPingScheduler.kt

```diff
@@ -0,0 +1,99 @@
+                .setPersisted(true)
+                .setRequiredNetworkType(JobInfo.NETWORK_TYPE_ANY)
+                .build()
+
+            // Get an instance of the JobScheduler from the application context
+            val jobScheduler =
+                applicationContext.getSystemService(Context.JOB_SCHEDULER_SERVICE) as JobScheduler
+
+            // Check to see if the periodic schedule is already started and start it if needed.
+            // Since the job is set up to be persisted, we may not need to start it again even
+            // after a reboot, which is why we are checking the current scheduled jobs for an
+            // instance that matches the job ID of the metrics ping.
+            if (!jobScheduler.allPendingJobs.contains(metricsPingJobInfo)) {
+                // Apply metrics ping default schedule.
+                jobScheduler.schedule(metricsPingJobInfo)
+            }
+        }
+    }
+
+    /**
+     * This method is called if the system has determined that you must stop execution of your job
+     * even before you've had a chance to call [.jobFinished].
+     *
+     * This will happen if the requirements specified at schedule time are no longer met. For
+     * example you may have requested WiFi with
+     * [android.app.job.JobInfo.Builder.setRequiredNetworkType], yet while your
+     * job was executing the user toggled WiFi. Another example is if you had specified
+     * [android.app.job.JobInfo.Builder.setRequiresDeviceIdle], and the phone left its
+     * idle maintenance window. You are solely responsible for the behavior of your application
+     * upon receipt of this message; your app will likely start to misbehave if you ignore it.
+     *
+     * Once this method returns, the system releases the wakelock that it is holding on
+     * behalf of the job.
+     *
+     * @param params The parameters identifying this job, as supplied to the job in the
+     *        [.onStartJob] callback.
+     * @return `true` to indicate to the JobManager whether you'd like to reschedule this job based
+     *         on the retry criteria provided at job creation-time; or `false` to end the job
+     *         entirely.  Regardless of the value returned, your job must stop executing.
+     */
+    override fun onStopJob(params: JobParameters?): Boolean {
+        // Use the Job handle to cancel the async task
+        metricsPingJob?.cancel()
+        return true
+    }
+
+    /**
+     * Called by the [JobScheduler] to asynchronously send the metrics ping and then calls
+     * [.jobFinished] to release the wake lock.
+     *
+     * @param params Parameters specifying info about this job
+     */
+    override fun onStartJob(params: JobParameters?): Boolean {
+        // handleEvent() should return a Job which we will need in order to interact with the
+        // async task uploading the ping
+        metricsPingJob = Glean.handleEvent(Glean.PingEvent.Default)
```

Comment:
Mh, so, I'm not sure how would this work: does this only work when the application is on and in the foreground? If it does work when the application is closed as well, what kind of metrics will it contain?

@pocmo have you ever used the job scheduler? What's your thought on this?

Reply:
Yes, we have been using `JobScheduler` in the old telemetry library. For all new code please use `WorkManager` from androix:
* https://developer.android.com/reference/androidx/work/WorkManager
* https://developer.android.com/topic/libraries/architecture/workmanager/

It uses `JobScheduler` and other mechanisms under the hood and avoids various issues.

Anyhow, regarding your question: At some point you schedule a job with some constraints (e.g. "I need network") (see `schedule` method here and JobScheduler will run your task in the background at some point (exceture `JobService`).

In our old telemetry we do the following: When we want to upload a ping, we generate the ping and write it to disk, we then schedule a background job (via JobScheduler) that will look at the stored pings on disk and try to upload them.

Is that helpful? Happy to write more :)

Reply:
Ah, this is very, very helpful, thanks!

I think we should scope this PR down and retain the same behaviour we have for other pings:

- check if some hours have passed since the last ping upload (stick to 24hrs for now)
- try to send the ping.

We have [bug 1508965](https://bugzilla.mozilla.org/show_bug.cgi?id=1508965) for using the `WorkManager`. We should probably do as the old telemetry was doing: serialize to disk and upload using the scheduler.

Reply:
I’m still waiting on final decisions about when the ping should be scheduled so I started this in order to start getting feedback.  Work manager sounds like a better API so will go that route.  JobScheduler here should work regardless of whether the app is in the foreground and should persist over device reboots, sending on a period of no more than once a day.  It will contain everything in the “default” ping or any metric that doesn’t specify a ping, as the spec indicated (unless I misread what the metrics ping was to send).

Since our ping data is already serialized to disk in SharedPreferences this JobScheduler using the provided JobInfo just triggers the handler onStart() which builds the ping just like the baseline and events pings.  I’d be happy to go through my thinking on this to make sure I’m not overlooking anything on this.

Reply:
> I’m still waiting on final decisions about when the ping should be scheduled so I started this in order to start getting feedback. Work manager sounds like a better API so will go that route. JobScheduler here should work regardless of whether the app is in the foreground and should persist over device reboots, sending on a period of no more than once a day. It will contain everything in the “default” ping or any metric that doesn’t specify a ping, as the spec indicated (unless I misread what the metrics ping was to send).

Let's stick to sending this ping every 24hrs, retaining the behaviour that Focus has for now.
Did you check how this behaves if the application is closed and there are ping lifetime metrics in memory?

> Since our ping data is already serialized to disk in SharedPreferences this JobScheduler using the provided JobInfo just triggers the handler onStart() which builds the ping just like the baseline and events pings. I’d be happy to go through my thinking on this to make sure I’m not overlooking anything on this.

Most of the metrics that will live in the `metrics` ping will have `Ping` lifetime, so they will not be persisted to disk. Moreover, relying on an internal implementation detail to power the scheduling logic is something we shouldn't probably do.

I think eventually persisting the whole ping to disk in order to enable background upload, retrying and different upload policies is the right way to move forward-

Reply:
This does stick to the every 24 hours, just not at a specific time of day.  I went ahead with this per @georgf suggestion in our meeting last week while we wait on the data team to determine when the ping should be sent.

Unless I'm missing something, even the Application lifetime pings get written to SharedPreferences, just the lifetime indicates whether or not they get cleared out when they are collected, where User lifetime is not cleared.  So when the schedule is triggered by JobScheduler, the metrics ping data is collected, formatted, and then cleared if the lifetime requires it.  This isn't an internal implementation just relying on [JobScheduler](https://developer.android.com/reference/android/app/job/JobScheduler) API to trigger when the ping is collected and sent.  @Dexterp37 I'd be happy to go through this on Vidyo with you if that helps to explain what my thinking was on this, it seems like we have a bit of a disconnect on this.

Reply:
I was overlooking the line where glean wasn't writing to SharedPreferences based on the ping lifetime, so this will need a little re-engineering per my conversation with @Dexterp37 just now.  Going to go the simple route and refactor to get glean sending metrics ping ASAP.